packages/drivers/odsp-driver/src/epochTracker.ts

```diff
@@ -0,0 +1,156 @@
+            }
+            return value.value as T;
+        }
+    }
+
+    /**
+     * Api to fetch the response for given request and parse it as json.
+     * @param url - url of the request
+     * @param fetchOptions - fetch options for request containing body, headers etc.
+     * @param addInBody - Pass True if caller wants to add epoch in post body.
+     */
+    public async fetchAndParseAsJSON<T>(
+        url: string,
+        fetchOptions: {[index: string]: any},
+        addInBody: boolean = false,
+    ): Promise<IOdspResponse<T>> {
+        // Add epoch in fetch request.
+        const request = this.addEpochInRequest(url, fetchOptions, addInBody);
+        try {
+            const response = await fetchAndParseAsJSONHelper<T>(request.url, request.fetchOptions);
+            this.validateEpochFromResponse(response.headers.get("x-fluid-epoch"));
+            return response;
+        } catch (error) {
+            this.checkForEpochError(error);
+            throw error;
+        }
+    }
+
+    /**
+     * Api to fetch the response as it is for given request.
+     * @param url - url of the request
+     * @param fetchOptions - fetch options for request containing body, headers etc.
+     * @param addInBody - Pass True if caller wants to add epoch in post body.
+     */
+    public async fetchResponse(
+        url: string,
+        fetchOptions: {[index: string]: any},
+        addInBody: boolean = false,
+    ): Promise<Response> {
+        // Add epoch in fetch request.
+        const request = this.addEpochInRequest(url, fetchOptions, addInBody);
+        try {
+            const response = await fetchHelper(request.url, request.fetchOptions);
+            this.validateEpochFromResponse(response.headers.get("x-fluid-epoch"));
+            return response;
+        } catch (error) {
+            this.checkForEpochError(error);
+            throw error;
+        }
+    }
+
+    private addEpochInRequest(
+        url: string,
+        fetchOptions: {[index: string]: any},
+        addInBody: boolean): {url: string, fetchOptions: {[index: string]: any}} {
+        if (this.fluidEpoch !== undefined) {
+            if (addInBody) {
+                // We use multi part form request for post body where we want to use this.
+                // So extract the form boundary to mark the end of form.
+                let body: string = fetchOptions.body;
+                const formBoundary = body.split("\r\n")[0].substring(2);
+                body += `\r\nepoch=${this.fluidEpoch}\r\n`;
+                body += `\r\n--${formBoundary}--`;
+                fetchOptions.body = body;
+            } else {
+                const [mainUrl, queryString] = url.split("?");
+                const searchParams = new URLSearchParams(queryString);
+                searchParams.append("epoch", this.fluidEpoch);
+                const urlWithEpoch = `${mainUrl}?${searchParams.toString()}`;
+                if (urlWithEpoch.length > 2048) {
```

Comment:
Can you please refactor code in getUrlAndHeadersWithAuth() and use logic in there here, such that there is only one place in code where 2048 number is encoded? Such that if we need to change it in future, we do not need to search for it everywhere in code?

Reply:
Same as my comment: https://github.com/microsoft/FluidFramework/pull/3785#discussion_r512180379. I will do a follow up on my PR for this. As this will end up in the class where we will add url, headers, query params etc and then pack them.